packages/network-controller/src/create-auto-managed-network-client.ts

```diff
@@ -0,0 +1,172 @@
+import { createNetworkClient, NetworkClient } from './create-network-client';
+import { BlockTracker, NetworkClientConfiguration, Provider } from './types';
+
+/**
+ * Represents a proxy object which wraps a target object. As a proxy, it allows
+ * for accessing and setting all of the properties that the target object
+ * supports, but also supports an extra propertyName (`__target__`) to access
+ * the target itself.
+ *
+ * @template Type - The type of the target object. It is assumed that this type
+ * will be constant even when the target is swapped.
+ */
+export type ProxyWithAccessibleTarget<TargetType> = TargetType & {
+  __target__: TargetType;
+};
+
+/**
+ * An object that provides the same interface as a network client but where the
+ * network client is not initialized until either the provider or block tracker
+ * is first accessed.
+ */
+export type AutoManagedNetworkClient<
+  Configuration extends NetworkClientConfiguration,
+> = {
+  configuration: Configuration;
+  provider: ProxyWithAccessibleTarget<Provider>;
+  blockTracker: ProxyWithAccessibleTarget<BlockTracker>;
+  destroy: () => void;
+};
+
+/**
+ * By default, the provider and block provider proxies will point to nothing.
+ * This is impossible when using the Proxy API, as the target object has to be
+ * something, so this object represents that "something".
+ */
+const UNINITIALIZED_TARGET = { __UNINITIALIZED__: true };
+
+/**
+ * The name of the method on both the provider and block tracker proxy which can
+ * be used to get the underlying provider or block tracker from the network
+ * client, when it is initialized.
+ */
+const REFLECTIVE_PROPERTY_NAME = '__target__';
+
+/**
+ * This function creates two proxies, one that wraps a provider and another that
+ * wraps a block tracker. These proxies are unique in that both will be "empty"
+ * at first; that is, neither will point to a functional provider or block
+ * tracker. They are set up in this way so as not to create unexpected network
+ * requests if the consumer of the proxy never accesses the network. Instead, as
+ * soon as a method or event is accessed on either object that requires a
+ * network request to function, a network client is created on the fly and
+ * the method or event in question is then forwarded to whichever part of the
+ * network client is serving as the receiver. The network client is then cached
+ * for subsequent usages. It can be destroyed at any time if it is no longer
+ * needed.
+ *
+ * @param networkClientConfiguration - The configuration object that will be
+ * used to instantiate the network client when it is needed.
+ * @returns The auto-managed network client.
+ */
+export function createAutoManagedNetworkClient<
```

Comment:
Do you think we should initialize these network clients upon construction as well? e.g. do something like `lookupNetwork`.

If we don't then we're in this strange situation where the selected network gets nicer error handling (inform the user when blocked, inform the UI when the network appears to be down) and populates more network metadata (EIP-1559 compatibility and network ID).

Reply:
This would need pretty significant state changes though. Maybe this can wait for another PR.

Reply:
Yeah, maybe I think about this along with making this step lazy.